conn.go

```diff
@@ -178,41 +194,41 @@ func NewConnWith(conn net.Conn, config ConnConfig) *Conn {
 			}},
 		}},
 	}).size()
-	c.selectVersions()
 	c.fetchMaxBytes = math.MaxInt32 - c.fetchMinSize
 	return c
 }
 
-func (c *Conn) selectVersions() {
-	var err error
-	apiVersions, err := c.ApiVersions()
+func (c *Conn) negotiateVersion(key apiKey, sortedSupportedVersions ...apiVersion) (apiVersion, error) {
+	v, err := c.loadVersions()
 	if err != nil {
-		c.apiVersions = defaultApiVersions
-	} else {
-		c.apiVersions = make(map[apiKey]ApiVersion)
-		for _, v := range apiVersions {
-			c.apiVersions[apiKey(v.ApiKey)] = v
-		}
+		return -1, err
 	}
-	for _, v := range c.apiVersions {
-		if apiKey(v.ApiKey) == fetchRequest {
-			switch version := v.MaxVersion; {
-			case version >= 10:
-				c.fetchVersion = 10
-			case version >= 5:
-				c.fetchVersion = 5
-			default:
-				c.fetchVersion = 2
-			}
-		}
-		if apiKey(v.ApiKey) == produceRequest {
-			if v.MaxVersion >= 7 {
-				c.produceVersion = 7
-			} else {
-				c.produceVersion = 2
-			}
-		}
+	a := v.negotiate(key, sortedSupportedVersions...)
+	if a < 0 {
+		return -1, fmt.Errorf("no matching versions were found between the client and the broker for API key %d", key)
 	}
+	return a, nil
+}
+
+func (c *Conn) loadVersions() (apiVersions, error) {
+	v, _ := c.apiVersions.Load().(apiVersions)
+	if v != nil {
+		return v, nil
+	}
+
+	brokerVersions, err := c.ApiVersions()
+	if err != nil {
+		return nil, err
+	}
+
+	v = make(apiVersions, len(brokerVersions))
+
+	for _, a := range brokerVersions {
+		v[apiKey(a.ApiKey)] = a
+	}
+
+	c.apiVersions.Store(v)
```

Comment:
I'm trying to determine whether there are any cases where we'd run into the thundering herd problem here.  The returned API versions will be consistent (at least I'd assume the broker we're connected to isn't changing its API versions out from under us!), but we could end up firing the call many times.  Do we expect much in the way of parallel produces or fetches?  I also am curious whether, as we add more uses cases for API negotiation, the thundering herd could become more an issue.  WDYT?

Reply:
I think the use cases we have right now are mostly single inflight requests, but you make a good point that the API is safe to use concurrently from multiple goroutines.

I can change it to use a blocking model so we guarantee that only one API Version request will be sent.

Reply:
On second thoughts, the number of API Version requests will never exceed the number of concurrent Produce or Fetch requests being sent, so while it may double the number of round trips it is strictly bound, there's no risk of exponentially increasing the request volume.

Let me know if you think that's good enough of a guarantee, otherwise I can change the synchronization mechanism.

Reply:
That sounds fine for now.  We can always revisit in the future if/when we find ourselves doing a lot more API version negotiation.  I think concurrent requests are also a pretty rare case on a single connection.